File: web/src/check.ts
import ArraySchema from "./types/ArraySchema";
import CheckError from "./types/CheckError";
import CheckStatus from "./types/CheckStatus";
import DefSchema from "./types/DefSchema";
import FieldSchema from "./types/FieldSchema";
import MixSchema from "./types/MixSchema";
import ObjectSchema from "./types/ObjectSchema";
import PropsSchema from "./types/PropsSchema";
import RefSchema from "./types/RefSchema";
import Schema from "./types/Schema";
import SchemaValue from "./types/SchemaValue";
import UnionSchema from "./types/UnionSchema";
import { createRegex } from "./utils";
import validators from "./validators";

interface CheckSuccess {
	ok: true;
}

interface CheckFailure {
	ok: false;
	errors: CheckError[];
}

export default function check(
	input: Record<PropertyKey, unknown>,
	schema: Schema,
): CheckSuccess | CheckFailure {
	let status: CheckStatus = {
		path: [],
		errors: [],
		defs: {},
	};

	checkObjectSchemaInner(input, schema, status);

	if (status.errors.length === 0) {
		return { ok: true };
	} else {
		return { ok: false, errors: status.errors };
	}
}

function checkObjectSchema(
	input: Record<PropertyKey, unknown>,
	schema: ObjectSchema,
	status: CheckStatus,
): boolean {
	return checkObjectSchemaInner(input, schema.inner, status);
}

function checkObjectSchemaInner(
	input: Record<PropertyKey, unknown>,
	schema: Schema,
	status: CheckStatus,
): boolean {
	let result = true;
	for (let [field, fieldSchema] of Object.entries(schema)) {
		status.path.push(field);
		if (field.startsWith("def$")) {
			// Add it to the status for use when it's referenced
			status.defs[(fieldSchema as DefSchema).name] = (fieldSchema as DefSchema).inner;
		} else if (field.startsWith("ref$")) {
			if (!checkRefSchema(input, fieldSchema as RefSchema, status)) {
				result = false;
			}
		} else if (field.startsWith("mix$")) {
			if (!checkMixSchema(input, fieldSchema as MixSchema, status)) {
				result = false;
			}
		} else if (field.startsWith("props$")) {
			if (!checkPropsSchema(input, fieldSchema as PropsSchema, field, status)) {
				result = false;
			}
		} else {
			const value = input[field];
			if (!checkFieldSchema(value, fieldSchema, field, status)) {
				result = false;
			}
		}
		status.path.pop();
	}
	return result;
}

function checkArraySchema(
	input: Record<PropertyKey, unknown>[],
	schema: ArraySchema,
	status: CheckStatus,
): boolean {
	let result = true;
	for (let [i, value] of input.entries()) {
		status.path.push(String(i));
		if (!checkFieldSchema(value, schema.inner, i.toString(), status)) {
			result = false;
		}
		status.path.pop();
	}
	return result;
}

function checkUnionSchema(value: unknown, schema: UnionSchema, field: string, status: CheckStatus) {
	let fieldStatus: CheckStatus = {
		path: [...status.path],
		errors: [],
		defs: status.defs,
	};
	let ok = false;
	for (let fs of schema.inner) {
		if (checkFieldSchema(value, fs, field, fieldStatus)) {
			ok = true;
			break;
		}
	}
	if (!ok) {
		status.errors.push({
			path: [...status.path],
			message: fieldStatus.errors.map((e) => e.message).join(" | "),
		});
	}
	return ok;
}

function checkRefSchema(input: Record<PropertyKey, unknown>, ref: RefSchema, status: CheckStatus) {
	let def = status.defs[ref.inner];
	if (def === undefined) {
		// This should never happen...
		status.errors.push({
			path: [...status.path],
			message: `Undefined def: ${ref}`,
		});
		return false;
	}

	return checkObjectSchemaInner(input, def, status);
}

function checkMixSchema(
	input: Record<PropertyKey, unknown>,
	schema: MixSchema,
	status: CheckStatus,
) {
	/*
	let fieldErrors: CheckError[] = [];
	let ok = false;
	for (let fs of schema.inner) {
		const mixResult = check(input, fs);
		if (mixResult.ok) {
			ok = true;
			break;
		} else {
			fieldErrors.push({
				path: [...status.path],
				message: mixResult.errors.map((e) => e.message).join(" & "),
			});
		}
	}
	if (!ok) {
		status.errors.push({
			path: [...status.path],
			message: fieldErrors.map((e) => e.message).join(" | "),
		});
	}
	return ok;
	*/
	let fieldErrors: CheckError[] = [];
	let ok = false;
	for (let fs of schema.inner) {
		let fieldStatus: CheckStatus = {
			path: [...status.path],
			errors: [],
			defs: status.defs,
		};
		if (checkObjectSchemaInner(input, fs, fieldStatus)) {
			ok = true;
			break;
		} else {
			fieldErrors.push({
				path: [...status.path],
				message: fieldStatus.errors.map((e) => e.message).join(" & "),
			});
		}
	}
	if (!ok) {
		status.errors.push({
			path: [...status.path],
			message: fieldErrors.map((e) => e.message).join(" | "),
		});
	}
	return ok;
}

function checkPropsSchema(
	input: Record<PropertyKey, unknown>,
	schema: PropsSchema,
	field: string,
	status: CheckStatus,
) {
	let result = true;
	for (let [anyField, value] of Object.entries(input)) {
		if (schema.type) {
			// PERF: could cache this
			const regexp = createRegex(schema.type);
			if (regexp === undefined) {
				// This should never happen...
				status.errors.push({
					path: [...status.path],
					message: `Unsupported pattern for '${field}': ${schema.type}`,
				});
				return false;
			}
			if (!regexp.test(anyField)) {
				status.errors.push({
					path: [...status.path],
					message: `'${anyField}' name doesn't match pattern '${schema.type}'`,
				});
				return false;
			}
		}

		// Run the field's validators
		if (!checkFieldSchema(value, schema.inner, anyField, status)) {
			result = false;
		}
	}
	return result;
}

function checkFieldSchema(
	value: unknown,
	schema: SchemaValue,
	field: string,
	status: CheckStatus,
): boolean {
	if (schema.type === "object") {
		if (value === null || typeof value !== "object") {
			status.errors.push({
				path: [...status.path],
				message: `'${field}' must be an object`,
			});
			return false;
		}
		return checkObjectSchema(value as Record<PropertyKey, unknown>, schema as ObjectSchema, status);
	} else if (schema.type === "array") {
		if (!Array.isArray(value)) {
			status.errors.push({
				path: [...status.path],
				message: `'${field}' must be an array`,
			});
			return false;
		}
		return checkArraySchema(value, schema as ArraySchema, status);
	} else if (schema.type === "union") {
		return checkUnionSchema(value, schema as UnionSchema, field, status);
	} else {
		return checkFieldSchemaValue(value, schema, field, status);
	}
}

function checkFieldSchemaValue(
	value: unknown,
	schema: FieldSchema,
	field: string,
	status: CheckStatus,
): boolean {
	if (value === undefined && schema.type !== "undef") {
		status.errors.push({
			path: [...status.path],
			message: `Field not found: ${field}`,
		});
		return false;
	}

	// Check the value's type
	switch (schema.type) {
		case "undef":
			if (value !== undefined) {
				status.errors.push({
					path: [...status.path],
					message: `'${field}' must be undefined`,
				});
				return false;
			}
			break;
		case "bool":
			if (typeof value !== "boolean") {
				status.errors.push({
					path: [...status.path],
					message: `'${field}' must be a boolean value`,
				});
				return false;
			}
			break;
		case "int":
			if (typeof value !== "number" || !Number.isInteger(value)) {
				status.errors.push({
					path: [...status.path],
					message: `'${field}' must be an integer value`,
				});
				return false;
			}
			break;
		case "num":
			if (typeof value !== "number") {
				status.errors.push({
					path: [...status.path],
					message: `'${field}' must be a number value`,
				});
				return false;
			}
			break;
		case "date":
			if (value instanceof Date === false) {
				status.errors.push({
					path: [...status.path],
					message: `'${field}' must be a date value`,
				});
				return false;
			}
			break;
		case "string":
			if (typeof value !== "string") {
				status.errors.push({
					path: [...status.path],
					message: `'${field}' must be a string value`,
				});
				return false;
			}
			break;
		default:
			// It may be a fixed value e.g. `true`
			let expectedType = schema.type;
			if (expectedType.startsWith('"') && expectedType.endsWith('"')) {
				expectedType = expectedType.substring(1, expectedType.length - 1);
			}
			if (String(expectedType) !== String(value)) {
				status.errors.push({
					path: [...status.path],
					message: `'${field}' must be '${expectedType}'`,
				});
				return false;
			}

			// There can't be any more validators after a fixed value
			return true;
	}

	// Run the validators
	if (schema.validators !== undefined) {
		for (let [method, { raw, required }] of Object.entries(schema.validators)) {
			if (method === "type" || method === "description") {
				continue;
			}

			const validate = validators[schema.type][method];
			if (validate !== undefined) {
				if (!validate(field, value, raw, required, status)) {
					return false;
				}
			} else {
				// This should never happen...
				status.errors.push({
					path: [...status.path],
					message: `Unsupported validation method for '${field}': ${method}`,
				});
				return false;
			}
		}
	}

	return true;
}

File: swift/Sources/sdnx/Check.swift
import Foundation
import Collections

// MARK: - Check Result Types

public struct CheckSuccess {
    public let ok = true
}

public struct CheckFailure {
    public let ok = false
    public let errors: [CheckError]
}

public enum CheckResult {
    case success(CheckSuccess)
    case failure(CheckFailure)
}

// MARK: - Main Check Function

public func check(_ input: OrderedDictionary<String, Any>, schema: Schema) -> CheckResult {
    var status = CheckStatus(path: [], errors: [], defs: [:])
    
    _ = checkObjectSchemaInner(input, schema: schema, status: &status)
    
    if status.errors.isEmpty {
        return .success(CheckSuccess())
    } else {
        return .failure(CheckFailure(errors: status.errors))
    }
}

// MARK: - Check Functions

func checkObjectSchema(_ input: OrderedDictionary<String, Any>, schema: ObjectSchema, status: inout CheckStatus) -> Bool {
    return checkObjectSchemaInner(input, schema: schema.inner, status: &status)
}

func checkObjectSchemaInner(_ input: OrderedDictionary<String, Any>, schema: Schema, status: inout CheckStatus) -> Bool {
    var result = true
    for (field, fieldSchema) in schema {
        status.path.append(field)
        if field.hasPrefix("def$") {
            if let defSchema = fieldSchema as? DefSchema {
                status.defs[defSchema.name] = defSchema.inner
            }
        } else if field.hasPrefix("ref$") {
            if let refSchema = fieldSchema as? RefSchema {
                if !checkRefSchema(input, ref: refSchema.inner, status: &status) {
                    result = false
                }
            }
        } else if field.hasPrefix("mix$") {
            if let mixSchema = fieldSchema as? MixSchema {
                if !checkMixSchema(input, schema: mixSchema, status: &status) {
                    result = false
                }
            }
        } else if field.hasPrefix("props$") {
            if let propsSchema = fieldSchema as? PropsSchema {
                if !checkPropsSchema(input, schema: propsSchema, field: field, status: &status) {
                    result = false
                }
            }
        } else {
            let value = input[field]
            if !checkFieldSchema(value, schema: fieldSchema, field: field, status: &status) {
                result = false
            }
        }
        status.path.removeLast()
    }
    return result
}

func checkArraySchema(_ input: [Any], schema: ArraySchema, status: inout CheckStatus) -> Bool {
    var result = true
    for (i, value) in input.enumerated() {
        status.path.append(String(i))
        if !checkFieldSchema(value, schema: schema.inner, field: String(i), status: &status) {
            result = false
        }
        let _ = status.path.popLast()
    }
    return result
}

func checkUnionSchema(_ value: Any?, schema: UnionSchema, field: String, status: inout CheckStatus) -> Bool {
    var fieldStatus = CheckStatus(path: status.path, errors: [], defs: status.defs)
    var ok = false
    for fs in schema.inner {
        if checkFieldSchema(value, schema: fs, field: field, status: &fieldStatus) {
            ok = true
            break
        }
    }
    if !ok {
        status.errors.append(CheckError(
            path: status.path,
            message: fieldStatus.errors.map { $0.message }.joined(separator: " | ")
        ))
    }
    return ok
}

func checkRefSchema(_ input: OrderedDictionary<String, Any>, ref: String, status: inout CheckStatus) -> Bool {
    guard let def = status.defs[ref] else {
        status.errors.append(CheckError(
            path: status.path,
            message: "Undefined def: \(ref)"
        ))
        return false
    }
    
    return checkObjectSchemaInner(input, schema: def, status: &status)
}

func checkMixSchema(_ input: OrderedDictionary<String, Any>, schema: MixSchema, status: inout CheckStatus) -> Bool {
    var fieldErrors: [CheckError] = []
    var ok = false
    for fs in schema.inner {
        var fieldStatus = CheckStatus(path: status.path, errors: [], defs: status.defs)
        if checkObjectSchemaInner(input, schema: fs, status: &fieldStatus) {
            ok = true
            break
        } else {
            fieldErrors.append(CheckError(
                path: status.path,
                message: fieldStatus.errors.map { $0.message }.joined(separator: " & ")
            ))
        }
    }
    if !ok {
        status.errors.append(CheckError(
            path: status.path,
            message: fieldErrors.map { $0.message }.joined(separator: " | ")
        ))
    }
    return ok
}

func checkPropsSchema(_ input: OrderedDictionary<String, Any>, schema: PropsSchema, field: String, status: inout CheckStatus) -> Bool {
    var result = true
    for (anyField, value) in input {
        // PERF: could cache this
        if let regexp = createRegex(schema.type) {
            if !regexp.test(anyField) {
                status.errors.append(CheckError(
                    path: status.path,
                    message: "'\(anyField)' name doesn't match pattern '\(schema.type)'"
                ))
                return false
            }
        }
        
        // Run the field's validators
        if !checkFieldSchema(value, schema: schema.inner, field: anyField, status: &status) {
            result = false
        }
    }
    return result
}

func checkFieldSchema(_ value: Any?, schema: SchemaValue, field: String, status: inout CheckStatus) -> Bool {
    switch schema.type {
    case "object":
        if let objSchema = schema as? ObjectSchema {
            guard let dictValue = value as? OrderedDictionary<String, Any> else {
                status.errors.append(CheckError(
                    path: status.path,
                    message: "'\(field)' must be an object"
                ))
                return false
            }
            return checkObjectSchema(dictValue, schema: objSchema, status: &status)
        }
        return false
        
    case "array":
        if let arrSchema = schema as? ArraySchema {
            guard let arrayValue = value as? [Any] else {
                status.errors.append(CheckError(
                    path: status.path,
                    message: "'\(field)' must be an array"
                ))
                return false
            }
            return checkArraySchema(arrayValue, schema: arrSchema, status: &status)
        }
        return false
        
    case "union":
        if let unionSchema = schema as? UnionSchema {
            return checkUnionSchema(value, schema: unionSchema, field: field, status: &status)
        }
        return false
        
    default:
        return checkFieldSchemaValue(value, schema: schema, field: field, status: &status)
    }
}

func checkFieldSchemaValue(_ value: Any?, schema: SchemaValue, field: String, status: inout CheckStatus) -> Bool {
    // Check if value is undefined (nil in Swift)
    // Allow "undef" and "null" types to have nil values
    if value == nil && schema.type != "undef" && schema.type != "null" {
        status.errors.append(CheckError(
            path: status.path,
            message: "Field not found: \(field)"
        ))
        return false
    }
    
    // Check the value's type
    switch schema.type {
    case "undef":
        if value != nil {
            status.errors.append(CheckError(
                path: status.path,
                message: "'\(field)' must be undefined"
            ))
            return false
        }
        
    case "null":
        if value != nil {
            // HACK: Can't figure out how to wrangle Swift nils
            guard value is String, value as? String == "null" else { 
                status.errors.append(CheckError(
                    path: status.path,
                    message: "'\(field)' must be null"
                ))
                return false
            }
        }
        
    case "bool":
        guard value is Bool else {
            status.errors.append(CheckError(
                path: status.path,
                message: "'\(field)' must be a boolean value"
            ))
            return false
        }
        
    case "int":
        guard let numValue = value as? NSNumber, numValue.isInt else {
            status.errors.append(CheckError(
                path: status.path,
                message: "'\(field)' must be an integer value"
            ))
            return false
        }
        
    case "num":
        guard let _ = value as? NSNumber else {
            status.errors.append(CheckError(
                path: status.path,
                message: "'\(field)' must be a number value"
            ))
            return false
        }
        
    case "date":
        guard value is Date else {
            status.errors.append(CheckError(
                path: status.path,
                message: "'\(field)' must be a date value"
            ))
            return false
        }
        
    case "string":
        guard let _ = value as? String else {
            status.errors.append(CheckError(
                path: status.path,
                message: "'\(field)' must be a string value"
            ))
            return false
        }
        
    default:
        // It may be a fixed value e.g. `true`, `false`, `null`, or a string
        var expectedType = schema.type
        if expectedType.hasPrefix("\"") && expectedType.hasSuffix("\"") {
            expectedType = String(expectedType.dropFirst().dropLast())
        }
        
        // Handle special cases for fixed values
        var matches = false
        if expectedType == "true" {
            matches = value as? Bool == true
        } else if expectedType == "false" {
            matches = value as? Bool == false
        } else if expectedType == "null" {
            matches = value == nil
        } else {
            // For string values, compare the string representation
            matches = value as? String == expectedType
        }
        
        if !matches {
            status.errors.append(CheckError(
                path: status.path,
                message: "'\(field)' must be '\(expectedType)'"
            ))
            return false
        }
        
        // There can't be any more validators after a fixed value
        return true
    }
    
    // Run the validators
    if let validatorDict = schema.validators {
        for (method, validator) in validatorDict {
            if method == "type" || method == "description" {
                continue
            }
            
            if let validate = getValidators()[schema.type]?[method] {
                if !validate(field, value ?? NSNull(), validator.raw, validator.required ?? NSNull(), &status) {
                    return false
                }
            } else {
                // This should never happen...
                status.errors.append(CheckError(
                    path: status.path,
                    message: "Unsupported validation method for '\(field)': \(method)"
                ))
                return false
            }
        }
    }
    
    return true
}
